lib/hsa/mcwamp_hsa.cpp

```diff
@@ -271,6 +272,207 @@ static unsigned extractBits(unsigned v, unsigned pos, unsigned w)
+                elfio reader;
+
+                static unsigned int iter = 0u;
+                if (reader.load(!iter++ ? self : info->dlpi_name)) {
+                    auto it = find_section_if(
+                        reader, [](const class section* x) {
+                        return x->get_type() == SHT_SYMTAB;
+                    });
+
+                    if (it) {
+                        const symbol_section_accessor symtab{reader, it};
+
+                        for (auto i = 0u; i != symtab.get_symbols_num(); ++i) {
+                            auto tmp = read_symbol(symtab, i);
+
+                            if (tmp.type == STT_OBJECT &&
+                                tmp.sect_idx != SHN_UNDEF) {
+                                r.emplace(
+                                    move(tmp.name),
+                                    make_pair(tmp.value, tmp.size));
+                            }
+                        }
+                    }
+                }
+
+                return 0;
+            }, nullptr);
+        });
+
+        return r;
+    }
+
+    inline
+    const std::vector<hsa_agent_t>& all_agents()
+    {
+        using namespace std;
+
+        static vector<hsa_agent_t> r;
+        static once_flag f;
+
+        call_once(f, []() {
+            for (auto&& acc : hc::accelerator::get_all()) {
+                if (acc.is_hsa_accelerator()) {
+                    r.push_back(
+                        *static_cast<hsa_agent_t*>(acc.get_hsa_agent()));
+                }
+            }
+        });
+
+        return r;
+    }
+
+    inline
+    void associate_code_object_symbols_with_host_allocation(
+        const ELFIO::elfio& reader,
+        const ELFIO::elfio& self_reader,
+        ELFIO::section* code_object_dynsym,
+        ELFIO::section* process_symtab,
+        hsa_agent_t agent,
+        hsa_executable_t executable)
+    {
+        using namespace ELFIO;
+        using namespace std;
+
+        if (!code_object_dynsym || !process_symtab) return;
+
+        const auto undefined_symbols = copy_names_of_undefined_symbols(
+            symbol_section_accessor{reader, code_object_dynsym});
+
+        for (auto&& x : undefined_symbols) {
+            using RAII_global =
+                unique_ptr<void, decltype(hsa_amd_memory_unlock)*>;
+            static unordered_map<string, RAII_global> globals;
+            static once_flag f;
+            call_once(f, [=]() { globals.reserve(symbol_addresses().size()); });
+
+            if (globals.find(x) == globals.cend()) {
+                const auto it1 = symbol_addresses().find(x);
+
+                if (it1 == symbol_addresses().cend()) {
+                    throw runtime_error{
+                        "Global symbol: " + x + " is undefined."};
+                }
+
+                if (globals.find(x) == globals.cend()) {
+                    static mutex mtx;
+                    lock_guard<mutex> lck{mtx};
```

Comment:
I presume this mutex is guarding `globals` for the insertion of `x` so shouldn't the lock comes before the 2nd `find(x)`?

Reply:
Not quite; `find` is `const`, therefore thread safe, doing a const read under a lock would be a pessimisation. If we really did not find it (i.e. nobody added it in between our second search) then we actually go and lock mtx. Since we already reserved the maximum possible amount of memory, the insertion would not invalidate any iterators so the return value of find or cend would not be rendered bogus by another thread doing an insertion. Bad things will happen if some detached thread somewhere actually tries to do this, since it may be running after main and static destruction have returned, but that's a different story altogether and a programmer error.

Reply:
Don't you need to prevent another thread inserting the same `x` between the 2nd `find(x)` and the lock?

Reply:
I was about to say that that is benign since it would only yield multiple calls to hsa_amd_memory_lock, but that's incorrect since the destruction of the temporary would call hsa_amd_memory_unlock on possibly the same, already inserted p. Thus you are correct; fixed it, thank you.